README.md
# Phone Manager API
[![Build Status](https://travis-ci.org/brunobcardoso/phone-manager.svg?branch=master)](https://travis-ci.org/brunobcardoso/phone-manager)
[![Coverage Status](https://coveralls.io/repos/github/brunobcardoso/phone-manager/badge.svg?branch=master)](https://coveralls.io/github/brunobcardoso/phone-manager)

A web application to manage calls records, built with DJango and Django REST framework.
It receives call detail records and calculates monthly bills for a given telephone number.
The webservice can be accessed through the [link](https://phonemanager.herokuapp.com/).

## Prerequisites
This project requires Python3.7 and [Pipenv](https://pipenv.readthedocs.io/en/latest/).

## Installing

Clone the project and change directory
```console
git clone https://github.com/brunobcardoso/phone-manager
cd phone-manager
```
Once you have installed pipenv, just run `make install` and you are ready to go:
```console
make install
```
The development server will be running at: [http://localhost:8000/](http://localhost:8000/)

## Running the tests
Tests:
```console
make test
```
Coverage:
```console
make coverage
```

## Admin Panel
```console
make admin
```
The command above creates a superuser named `admin` with the password `pass`.

Admin Panel: [http://localhost:8000/admin/](http://localhost:8000/admin/)

## Usage

### Create a Call Start Record
*  `POST` `http://localhost:8000/records`

**Example**

request:
```console
curl -X POST \
  http://localhost:8000/records \
  -H 'Content-Type: application/json' \
  -d '{                                      
   "type":"start",
   "timestamp":"2018-01-10T21:50:13Z",
   "call_id":"101",
   "source":"1145678901",
   "destination":"11987654321"
}'
```

response:
```console
{
    "call_id": 101,
    "source": "1145678901",
    "destination": "11987654321",
    "type": "start",
    "timestamp": "2018-01-10T21:50:13Z"
}
```

### Create a Call End Record 
*  `POST` `http://localhost:8000/records`

**Example**

request:
```console
curl -X POST \
  http://localhost:8000/records \
  -H 'Content-Type: application/json' \
  -d '{                                      
   "type":"end",
   "timestamp":"2018-01-10T22:05:13Z",
   "call_id":"101"
}'
```

response:
```console
{
    "call_id": 101,
    "type": "end",
    "timestamp": "2018-01-10T22:05:13Z"
}
```

### Get telephone bill 

#### With just the subscriber telephone number
*  `GET` `http://localhost:8000/bills/<subscriber>`

**Example**

request:
```console
    curl -X GET http://localhost:8000/bills/1145678901 
```

response:
```console
{
    "subscriber": "1145678901",
    "reference_period": "08/2018",
    "bill_call_records": [
            {
            "destination": "11987654321",
            "call_start_date": "2018-08-28",
            "call_start_time": "13:29:33",
            "call_duration": "1h0m13s",
            "call_price": "R$ 5,76"
        }
    ]
}
``` 

#### With subscriber and reference period (month/year)
*  `GET` `http://localhost:8000/bills/<subscriber>?reference=MM/YYYY`

**Example**

request:

```console
curl -X GET 'http://localhost:8000/bills/1145678901?reference=01-2018'
```
response:

```console
{
    "subscriber": "1145678901",
    "reference_period": "01/2018",
    "bill_call_records": [
        {
            "destination": "11987654321",
            "call_start_date": "2018-01-10",
            "call_start_time": "21:50:13",
            "call_duration": "0h15m0s",
            "call_price": "R$ 1,26"
        },
        {
            "destination": "11987654321",
            "call_start_date": "2018-01-11",
            "call_start_time": "08:10:25",
            "call_duration": "0h45m0s",
            "call_price": "R$ 4,41"
        }
    ]
}
```

### Pricing Rules
The pricing related variables are stored at the project settings file
`phonemanager/settings.py`, with a default value as follow:

| Environment Variable  |             Description            | Default |
|-----------------------|:----------------------------------:|--------:|
| `STD_HOUR_START`      |      Standard time call start      |       6 |
| `STD_HOUR_END`        | Standard time call end (excluding) |      22 |
| `STD_STANDING_CHARGE` |      Standard Standing charge      |    0.36 |
| `STD_MINUTE_CHARGE`   |     Standard Call charge/minute    |    0.00 |
| `RDC_STANDING_CHARGE` |       Reduced Standing charge      |    0.36 |
| `RDC_MINUTE_CHARGE`   |     Reduced Call charge/minute     |    0.00 |

To change this variables for development purposes, set the variable in
`.env` file, and run the development server
```console
make run
```

## Entity Relationship Diagram(ERD)

![alt text](erd_core.png)

To keep the Database documentation synchronized with the source code, run `make erd`.
It will generate a graphical representation of app `core` models. 
For a complete ERD, run `make cerd` instead.
```console
make erd
```

## Deploy to Heroku
```console
heroku create
heroku addons:create heroku-postgresql:hobby-dev --app <your_app_name>
heroku plugins:install heroku-config
heroku config:set SECRET_KEY=<your_secret_key>
heroku config:set ALLOWED_HOSTS=['your_app_name.herokuapp.com’]
git push heroku master
heroku ps:scale web=1
```
To update any Pricing rule variable, run:
 ```console
heroku config:set <ENV_VARIABLE_NAME>=<new_value>
```
Or create an .env file with all then and just run `heroku config:push -a`
from the same directory:
 ```console
heroku config:push -a
```
It will send configs from .env to Heroku

## Work environment
|                       |                  |
|-----------------------|------------------|
| Computer              | Dell Vostro 5480 |
| S.O.                  | Linux Mint 19    |
| IDE                   | PyCharm          |
| Python                | 3.7.0            |
| Pipenv                | 2018.7.1         |
| Git-flow              | 1.11.0           |
| Django                | 2.1.1            |
| Django REST framework | 3.8.2            |


CHANGES.txt
Phone Manager Changelog
=======================

Version 0.1.6
-------------
- Add unique ongoing call validation
- Fix record overlapping

Version 0.1.5
-------------
- Fix phone number validator
- Fix a bug where StartRecordSerializer would cause an inconsistent state
partially saving only the call data if the record was invalid.

Version 0.1.4
-------------
- Improve readability on tests
- Keep validations in the model

Version 0.1.3
--------------
- Fix typo in README
- Add missing validation test unique_timestamp_for_destination

Version 0.1.2
--------------
- Fix ValidationError 500 'server error' response.
- Add a custom exception handling for DRF validation errors

Version 0.1.1
-------------

- Add validation for source and destination timestamp
- Validates if exists a call record for the same source/destination and timestamp

- Fix timestamp validation for endrecord
- Timestamp of end record cannot be equal to start record

- Fix standard_minutes calculator
- It was not considering the completed 60 seconds cycle and limits changes.

Version 0.1
-----------

- Initial release

phonemanager/settings.py
"""
Django settings for phonemanager project.

Generated by 'django-admin startproject' using Django 2.1.1.

For more information on this file, see
https://docs.djangoproject.com/en/2.1/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/2.1/ref/settings/
"""

import os

from decouple import config, Csv
from dj_database_url import parse as dburl

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/2.1/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = config('SECRET_KEY', default='secret_key')

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = config('DEBUG', default=False, cast=bool)

ALLOWED_HOSTS = config('ALLOWED_HOSTS', default='*', cast=Csv())

# Application definition

INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'django_extensions',
    'rest_framework',
    'core',
]

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
]

ROOT_URLCONF = 'phonemanager.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

WSGI_APPLICATION = 'phonemanager.wsgi.application'

# Database
# https://docs.djangoproject.com/en/2.1/ref/settings/#databases

default_dburl = 'sqlite:///' + os.path.join(BASE_DIR, 'db.sqlite3')
DATABASES = {
    'default': config(
        'DATABASE_URL',
        default=default_dburl,
        cast=dburl
    ),
}

# Password validation
# https://docs.djangoproject.com/en/2.1/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]

# Internationalization
# https://docs.djangoproject.com/en/2.1/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_L10N = True

USE_TZ = True

# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/2.1/howto/static-files/

STATIC_URL = '/static/'
STATIC_ROOT = os.path.join(BASE_DIR, 'staticfiles')

REST_FRAMEWORK = {
    'EXCEPTION_HANDLER': 'core.utils.custom_exception_handler'
}

"""
Phone Manager Pricing rules.
There are two tariff prices: Standard and Reduced
"""
# Standard Time
STD_HOUR_START = config('STD_HOUR_START', default=6, cast=int)
STD_HOUR_END = config('STD_HOUR_END', default=22, cast=int)
# Standard Prices
STD_STANDING_CHARGE = config('STD_STANDING_CHARGE', default=0.36)
STD_MINUTE_CHARGE = config('STD_MINUTE_CHARGE', default=0.09)

# Reduced Prices
RDC_STANDING_CHARGE = config('RDC_STANDING_CHARGE', default=0.36)
RDC_MINUTE_CHARGE = config('RDC_MINUTE_CHARGE', default=0)
